std/hash_map.zig

```diff
@@ -135,15 +135,40 @@ pub fn HashMap(comptime K: type, comptime V: type, comptime hash: fn (key: K) u3
             return res.kv;
         }
 
-        fn ensureCapacity(self: *Self) !void {
-            if (self.entries.len == 0) {
-                return self.initCapacity(16);
+        fn optimizedCapacity(expected_count: usize) usize {
+            // ensure that the hash map will be at most 60% full if
+            // new_capacity items are put into the hash map
+            var optimized_capacity = expected_count * 5 / 3;
+            // round capacity to the next power of two
+            const is_power_of_two = optimized_capacity & (optimized_capacity-1) == 0;
+            if (!is_power_of_two) {
+                const pow = math.log2_int_ceil(usize, optimized_capacity);
+                optimized_capacity = math.pow(usize, 2, pow);
             }
+            return optimized_capacity;
+        }
 
-            // if we get too full (60%), double the capacity
-            if (self.size * 5 >= self.entries.len * 3) {
-                const old_entries = self.entries;
-                try self.initCapacity(self.entries.len * 2);
+        /// Increase capacity so that the hash map will be at most
+        /// 60% full when expected_count items are put into it
+        pub fn ensureCapacity(self: *Self, expected_count: usize) !void {
+            const optimized_capacity = optimizedCapacity(expected_count);
+            return self.ensureCapacityExact(optimized_capacity);
+        }
+
+        /// Sets the capacity to the new capacity if the new
+        /// capacity is greater than the current capacity.
+        /// New capacity must be a power of two.
+        fn ensureCapacityExact(self: *Self, new_capacity: usize) !void {
+            const is_power_of_two = new_capacity & (new_capacity-1) == 0;
+            assert(is_power_of_two);
+
+            if (new_capacity <= self.entries.len) {
+                return;
+            }
+
+            const old_entries = self.entries;
+            try self.initCapacity(new_capacity);
```

Comment:
I think you might need a `.incrementModificationCount` around here somewhere.

Reply:
Previously, resizing didn't count as modification. Was that a bug? There was no `incrementModificationCount` call in `ensureCapacity` before.

Reply:
> Should the new public `ensureCapacity` call `incrementModificationCount`?

Only if it actually modifies the capacity.

Reply:
To be sure, this is because the following would/should fail?

```zig
    var it = map.iterator();
    var capacity: usize = 1024;
    while (it.next()) |next| {
        map.ensureCapacity(capacity);
        capacity = capacity * 2;
    }
```

Reply:
Yep. the code below (under "dump all of the old elements into the new table") will likely break the iterator state, and we want to be able to catch that.